pulp_2to3_migration/app/plugin/content.py

```diff
@@ -175,31 +193,71 @@ def __init__(self, migrator):
+            except ObjectDoesNotExist:
+                return
+            return pulp2importer.pulp3_remote
+
         for pulp2content in batch:
             pulp_2to3_detail_content = pulp2content.detail_model
             pulp3content = pulp_2to3_detail_content.create_pulp3_content()
             future_relations = {'pulp2content': pulp2content}
 
-            if not pulp2content.downloaded:
-                # on_demand content is partially migrated - only Content is created at this stage.
-                # Remote Artifact and Content Artifact should be created at the time of
-                # importers/remotes migration. Rely on downloaded flag on Pulp2Content to
-                # identify on_demand content.
-                dc = DeclarativeContent(content=pulp3content)
-            else:
-                artifact = await self.create_artifact(pulp2content.pulp2_storage_path,
-                                                      pulp_2to3_detail_content.expected_digests,
-                                                      pulp_2to3_detail_content.expected_size)
+            # get all Lazy Catalog Entries (LCEs) for this content
+            pulp2lazycatalog = Pulp2LazyCatalog.objects.filter(
+                pulp2_unit_id=pulp2content.pulp2_id)
+
+            # on_demand content can't be migrated if it has no LCE
+            if not pulp2lazycatalog and not pulp2content.downloaded:
+                break
+
+            artifact = await self.create_artifact(pulp2content.pulp2_storage_path,
+                                                  pulp_2to3_detail_content.expected_digests,
+                                                  pulp_2to3_detail_content.expected_size,
+                                                  downloaded=pulp2content.downloaded)
+
+            # Downloaded content with no LCE
+            if not pulp2lazycatalog and pulp2content.downloaded:
                 da = DeclarativeArtifact(
                     artifact=artifact,
                     url=NOT_USED,
                     relative_path=pulp_2to3_detail_content.relative_path_for_content_artifact,
-                    remote=NOT_USED,
+                    remote=None,
                     deferred_download=False)
                 dc = DeclarativeContent(content=pulp3content, d_artifacts=[da])
+                dc.extra_data = future_relations
+                await self.put(dc)
+
+            # Downloaded or on_demand content with LCEs.
+            #
+            # To create multiple remote artifacts, create multiple instances of declarative
+            # content which will differ by url/remote in their declarative artifacts
+            for lce in pulp2lazycatalog:
+                remote = get_remote_by_importer_id(lce.pulp2_importer_id)
+                deferred_download = not pulp2content.downloaded
+                if not remote and deferred_download:
+                    # on_demand content can't be migrated without a remote
```

Comment:
Should this create a warning for the user?

Reply:
Yeah, probably. +1